src/inmanta/server/services/orchestrationservice.py

```diff
@@ -238,24 +281,36 @@ def _merge_resources(
                         " be updated via a partial compile"
                     )
                 else:
-                    merged_resources[new_resource.resource.id] = new_resource.resource.dict()
-        return list(merged_resources.values())
+                    merged_resources[new_resource.get_resource_id_str()] = new_resource
+        return merged_resources
 
     def _merge_resource_sets(
-        self, old_resource_sets: Dict[ResourceIdStr, Optional[str]], paired_resources: List[PairedResource]
+        self, old_resource_sets: Dict[ResourceIdStr, Optional[str]], updated_resource_sets: Set[str]
     ) -> Dict[ResourceIdStr, Optional[str]]:
-        updated_resource_sets: Set[str] = set(
-            res.new_resource.resource_set for res in paired_resources if not res.new_resource.is_shared_resource()
-        )
         changed_resource_sets: Set[str] = updated_resource_sets.union(self.removed_resource_sets)
         unchanged_resource_sets: Dict[ResourceIdStr, Optional[str]] = {
             k: v for k, v in old_resource_sets.items() if v not in changed_resource_sets
         }
         return {**unchanged_resource_sets, **self.resource_sets}
 
-    async def apply_partial(
-        self, *, connection: asyncpg.connection.Connection
-    ) -> tuple[list[dict[str, object]], dict[ResourceIdStr, Optional[str]]]:
+    async def _merge_unknowns(
+        self, merged_resources: dict[ResourceIdStr, ResourceWithResourceSet]
+    ) -> List[data.UnknownParameter]:
+        """
+        Merge all relevant, unresolved unknowns from the old version of the model together with the unknowns
+        of the partial compile.
+        """
+        rids_in_partial_update = [resource_minimal.get_resource_id_str() for resource_minimal in self.partial_updates]
+        rids_not_in_partial_compile = [rid for rid in merged_resources if rid not in rids_in_partial_update]
```

Comment:
```suggestion
        rids_in_partial_update = {resource_minimal.get_resource_id_str() for resource_minimal in self.partial_updates}
        rids_not_in_partial_compile = {rid for rid in merged_resources if rid not in rids_in_partial_update}
```
There might be a great amount of resources (especially in the second list) so we better make the lookup efficient I think.